File: backend/pb_hooks/views/githubBotStarted.js
/// <reference path="../../pb_data/types.d.ts" />

/**
 * Represents the meta information for a task in JavaScript.
 * @typedef {Object} TaskMeta
 * @property {string} repository - The repository.
 * @property {string} run_id - The run ID.
 * @property {string} sha - The SHA value.
 */

/**
 * Represents the data structure of the request from github
 * @typedef {Object} GithubBody
 * @property {TaskMeta} meta - The meta information.
 */

routerAdd(
  "POST",
  "/github-bot-started",
  (c) => {
    const {
      findWorkspaceByRepository,
      findWorkspaceEventByWorkspace,
    } = require(`${__hooks}/utils/workspace.js`);

    /** @type {GithubBody} */
    const data = $apis.requestInfo(c).data;
    const workspace = findWorkspaceByRepository(data.meta.repository);
    const workspace_event = findWorkspaceEventByWorkspace(workspace);

    workspace_event.set("new_run_started", new Date());
    $app.dao().saveRecord(workspace_event);

    return c.json(201, {
      code: 201,
      message: "workflow event registered",
    });
  },
  $apis.activityLogger($app),
  $apis.requireGuestOnly()
);

File: backend/src/views/githubBotStarted.go
package views

import (
	"hackathon-backend/src/types"
	"hackathon-backend/src/utils"
	"net/http"
	"time"

	"github.com/labstack/echo/v5"
	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/apis"
	"github.com/pocketbase/pocketbase/core"
)

func AddGithubBotStarted(app *pocketbase.PocketBase, e *core.ServeEvent) {
	e.Router.AddRoute(echo.Route{
		Method: http.MethodPost,
		Path:   "/github-bot-started/",
		Handler: func(c echo.Context) error {
			var reqBody types.GithubRequestBody
			var err error

			if err = c.Bind(&reqBody); err != nil {
				return echo.NewHTTPError(http.StatusBadRequest, err)
			}

			workspace, err := utils.GetWorkspaceByMeta(app, &reqBody.Meta)
			if err != nil {
				return echo.NewHTTPError(http.StatusBadRequest, err)
			}

			workspace_event, err := utils.GetWorkspaceEventByWorkspace(app, workspace)
			if err != nil {
				return echo.NewHTTPError(http.StatusBadRequest, err)
			}

			now := time.Now()

			workspace_event.Set("new_run_started", now)

			if err := app.Dao().SaveRecord(workspace_event); err != nil {
				return err
			}

			resBody := &types.GithubResponseBody{
				Code:    201,
				Message: "workflow event registered",
			}

			return c.JSON(http.StatusCreated, resBody)
		},
		Middlewares: []echo.MiddlewareFunc{
			apis.ActivityLogger(app),
			apis.RequireGuestOnly(),
		},
	})
}